actors/builtin/reward/reward_state.go

```diff
@@ -49,31 +40,174 @@ func ConstructState(emptyMultiMapCid cid.Cid) *State {
+//    will always be between 0 and 1). The expectation is that callers will
+//    multiply the result by some number, and THEN right-shift the result of
+//    the multiplication by FixedPoint bits, thus implementing fixed-point
+//    multiplication by the returned fraction.
+//
+// 2. Since we do not have a math library in this setting, we cannot directly
+//    implement the intended closed form using stock implementations of
+//    elementary functions like exp and ln. Not even powf is available.
+//    Instead, we have manipulated the function into a form with a tractable
+//    Taylor expansion, and then implemented the fixed-point Taylor expansion
+//    in an efficient way.
+//
+// Mathematical Derivation
+//
+//   Note that the function f above is analytically equivalent to:
+//
+//                    (   ( 1 )              1                 )
+//                    ( ln( - ) * ------------------------ * t )
+//                    (   ( 2 )   [# of epochs in 6 years]     )
+//        f(t) = 1 - e
+//
+//   We define λ = -ln(1/2)/[# of epochs in 6 years]
+//               = -ln(1/2)*([Seconds per epoch] / (6 * [Seconds per year]))
+//   such that
+//                    -λt
+//        f(t) = 1 - e
+//
+//   Now, we substitute for the exponential its well-known Taylor series at 0:
+//
+//                 infinity     n
+//                   \```` (-λt)
+//        f(t) = 1 -  >    ------
+//                   /,,,,   n!
+//                   n = 0
+//
+//   Simplifying, and truncating to the empirically necessary precision:
+//
+//                  24           n
+//                \```` (-1)(-λt)
+//        f(t) =   >    ----------
+//                /,,,,     n!
+//                n = 1
+//
+//   This is the final mathematical form of what is computed below. What remains
+//   is to explain how the series calculation is carried out in fixed-point.
+//
+// Algorithm
+//
+//   The key observation is that each successive term can be represented as a
+//   rational number, and derived from the previous term by simple
+//   multiplications on the numerator and denominator. In particular:
+//   * the numerator is the previous numerator multiplied by (-λt)
+//   * the denominator is the previous denominator multiplied by n
+//   We also need to represent λ itself as a rational, so the denominator of
+//   the series term is actually multiplied by both n and the denominator of
+//   lambda.
+//
+//   When we have the numerator and denominator for a given term set up, we
+//   compute their fixed-point fraction by left-shifting the numerator before
+//   performing integer division.
+//
+//   Finally, at the end of each loop, we remove unnecessary bits of precision
+//   from both the numerator and denominator accumulators to control the
+//   computational complexity of the bigint multiplications.
+
+// Fixed-point precision used internally and for output
+const FixedPoint = 97
+
+// Used in the definition of λ
+const BlockTimeSeconds = 30
```

Comment:
FYI this number is wrong, and has a different value elsewhere in this repo. But that's tracked as a thing to fix (with a structural fix, not just changing the literal here)

Reply:
Yeah, I saw that mentioned in #317, and did some grepping around, but I was unable to find the other value.